src/OrbitVulkanLayer/SubmissionTracker.h

```diff
@@ -976,6 +893,140 @@ class SubmissionTracker : public VulkanLayerProducer::CaptureStatusListener {
+      ERROR_ONCE(
+          "Calling vkQueueSubmit on a command buffer that is in the initial state (i.e. "
+          "either freshly allocated or reset with vkResetCommandBuffer).");
+      return;
+    }
+    CHECK(command_buffer_to_state_.contains(command_buffer));
+    CommandBufferState& state = command_buffer_to_state_.at(command_buffer);
+    bool has_been_submitted_before = state.pre_submission_cpu_timestamp.has_value();
+
+    // Mark that this command buffer in the current state was already submitted. If the command
+    // buffer is being submitted again, without a "vkResetCommandBuffer" call, the same slot
+    // indices will be used again, so that we can not distinguish the results for the different
+    // submissions anymore. Thus, this field will be used to ensure to not try to read such a
+    // slot. Calling "vkResetCommandBuffer" (or "vkBeginCommandBuffer"), will reset this field.
+    // Note that we are using an std::optional with the submission time to protect us against an
+    // "OnCaptureFinished" call, right between pre and post submission, which would try to reset
+    // the slots.
+    state.pre_submission_cpu_timestamp =
+        queue_submission->meta_information.pre_submission_cpu_timestamp;
+
+    if (device == VK_NULL_HANDLE) {
+      device = command_buffer_to_device_.at(command_buffer);
+    }
+
+    // If we haven't recorded neither the end nor the begin of a command buffer, we have no
+    // information to send.
+    if (!state.command_buffer_end_slot_index.has_value()) {
+      if (state.command_buffer_begin_slot_index.has_value()) {
+        // We need to discard the begin slot when we only have a begin slot. This can happen
+        // if we run out of query slots.
+        query_slots_not_needed_to_read->push_back(state.command_buffer_begin_slot_index.value());
+        state.command_buffer_begin_slot_index.reset();
+      }
+      return;
+    }
+
+    // If this command buffer was already submitted before (without reset afterwards), its slot
+    // indices are not unique anymore and can't be used by us. Note, that the slots will be
+    // marked as "done with reading" when the first submission of this command buffer is done
+    // and will eventually be reset on a "vkResetCommandBuffer" call.
+    if (!has_been_submitted_before) {
+      SubmittedCommandBuffer submitted_command_buffer{
+          .command_buffer_begin_slot_index = state.command_buffer_begin_slot_index,
+          .command_buffer_end_slot_index = state.command_buffer_end_slot_index.value()};
+      submitted_submit_info->command_buffers.emplace_back(submitted_command_buffer);
+    }
+  }
+
+  void PersistDebugMarkersOfASingleCommandBufferOnSubmitUnsafe(
+      VkCommandBuffer command_buffer, std::optional<QueueSubmission>* queue_submission_optional,
+      QueueMarkerState* markers, std::vector<uint32_t>* marker_slots_not_needed_to_read) {
+    CHECK(queue_submission_optional != nullptr);
```

Comment:
Same as above, add `mutex_.AssertHeld();` and consider removing the "unsafe" in the name.

Reply:
Done.